Stm32h7 flash controller on m4

mkaricheri opened the pull request:
Hello,

We have been using STM32H7 based SoC in our project and made some updates to the flash controller driver on STMH7 SoC to allow it run on a M4 core. I am sending the updates from that work to benefit the SoC users.

The initial draft was discussed at https://github.com/zephyrproject-rtos/zephyr/pull/65250
Based on the feedback, I have removed the SoC flash example as I discovered that the flash shell command is sufficient to verify the flash functionality.

I have used the samples/drivers/flash_shell to test this patch by using the build commands below and using west flash to program the image onto the board.

west build -p always -b stm32h747i_disco/stm32h747xx/m4 -d m4 samples/drivers/flash_shell
west build -p always -b stm32h745i_disco/stm32h745xx/m4 -d m4 samples/drivers/flash_shell

I have an additional commits to allow testing which enable usart1 shell/console on M4 and disable the same for M7
https://github.com/mkaricheri/zephyr/tree/stm32h7_flash_controller_on_m4-test

I test flash erase/read/write as below for  stm32h747i_disco & stm32h745i_disco

```
*** Booting Zephyr OS build v3.7.0-rc2-66-g0d6dc6f64f4a ***
Flash shell sample
uart:~$ flash erase flash-controller@52002000 c0000 40000
Erase success.
[00:00:21.945,000] <err> flash_stm32h7: Cortex M4: ART enabled not supported
uart:~$ flash read flash-controller@52002000 c0000 32
000C0000: ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff |........ ........
000C0010: ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff |........ ........
000C0020: ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff |........ ........
000C0030: ff ff                                            |..
uart:~$ flash write flash-controller@52002000 c0000 12345678 23456789 3456789
Write OK.
Verified.
uart:~$ flash read flash-controller@52002000 c0000 32
000C0000: 78 56 34 12 89 67 45 23  90 78 56 34 ff ff ff ff |xV4..gE# .xV4....
000C0010: ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff |........ ........
000C0020: ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff |........ ........
000C0030: ff ff   
uart:~$ flash write flash-controller@52002000 c0000 12345678 23456789 3456789
Write internal ERROR!
uart:~$ flash page_info flash-controller@52002000 c0000
Page for address 0xc0000:
start offset: 0xc0000
size: 131072
index: 6

                              
```

mkaricheri committed:
drivers: flash: stm32h7: support flash controller driver on M4

Currently flash controller driver builds and runs only on M7.
This patch supports enablement of the driver on M4 CPUs. The
main issue in using the driver on M4 is that LL_GetFlashSize()
to read the flash size works only on M7 as the internal register
is not accessible from M4. So to use the driver on M4, add a dt
property, bank2-flash-size, to configure flash size of bank2.
this will allow gradual support of flash controller driver
on M4 of all supported STMH7 boards by defining the above
dt property and testing it. Currently this is verified only
on STM32H747i-disco board.

Signed-off-by: Murali Karicheri <murali.karicheri@sandc.com>

mkaricheri committed:
drivers: flash : stm32h7: Fix the offset for get_sector()

The offset is wrong when the second bank is used by M4 as the
offset is relative to the base which is 0. So add the least
significant 6 nibbles from the CONFIG_FLASH_BASE_ADDRESS to
find the sector.

Signed-off-by: Murali Karicheri <murali.karicheri@sandc.com>

mkaricheri commented:
Hello @erwango @FRASTM @ABESTM,

When you get a chance, could you please review this?

Thanks

Murali Karicheri

mkaricheri commented on dts/bindings/mtd/st,stm32-nv-flash.yaml:
```diff
@@ -12,3 +12,10 @@ properties:
+      This property defines the size in KB of the embedded flash memory bank2. This
+      is needed since M4 CPU to which bank2 is allocated doesn't have access to flash
+      controller register to read the size. This provides a way to configure this
+      size when the flash controller driver runs on the M4 CPU.
```
Updated. I have also rebased to latest on main and pushed

mkaricheri replied:
@ajarmouni-st while reviewing the 755 SoC PR from my colleague, I found I might have to remove 747i-disco board from [samples/drivers/flash_shell/sample.yaml ](https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/drivers/flash_shell/sample.yaml)

Will add that in a separate commit

mkaricheri replied:
I have updated the PR for the above and pushed.  Can be removed if not right,

mkaricheri committed:
samples: drivers: flash_shell: allow run on stm32h747i_disco/stm32h747xx/m4

Now that flash controller driver is supported on M4 of stm32h747i_disco,
remove it from platform_exclude list and add an overlay for flash_shell
sample

Signed-off-by: Murali Karicheri <murali.karicheri@sandc.com>

mkaricheri commented:
@erwango @ajarmouni-st do you think this requires additional review or rework or in a good shape to sail through to merge?

GeorgeCGV commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -36,11 +36,16 @@ LOG_MODULE_REGISTER(LOG_DOMAIN);
+#define STM32H7_M4_FLASH_SIZE DT_PROP_OR(DT_INST(0, st_stm32_nv_flash), bank2_flash_size, 0)
 #ifdef CONFIG_CPU_CORTEX_M4
+#if STM32H7_M4_FLASH_SIZE == 0
 #error Flash driver on M4 core is not supported yet
```
Shouldn't the error be adapted? As I understand, this PR adds the support but requires to set the size in DT.

mkaricheri replied:
Pushed the update @GeorgeCGV

FRASTM reviewed (commented):
LGTM, 
can you please give an example of the <bank2-flash-size:> property setting to access the flash from the CM4

mkaricheri commented:
> LGTM, can you please give an example of the bank2-flash-size: property setting to access the flash from the CM4

Okay the missing overlay for STM32H747i-disco_m4 is added in the last commit so that we could build and test sample/drivers/flash_shell sample for this board. Is it what you are looking for @FRASTM ?

mkaricheri commented:
@FRASTM @GeorgeCGV @ajarmouni-st CC:  @ABESTM, @erwango  I have addressed all of your comments and also rebased the branch to the latest on main and pushed. Please give your approval if this looks good.

FRASTM commented:
> @FRASTM @GeorgeCGV @ajarmouni-st CC: @ABESTM, @erwango I have addressed all of your comments and also rebased the branch to the latest on main and pushed. Please give your approval if this looks good.

I'd like please an example of the DTS for stm32h747_disco//m4 or stm32h745i_disco//m4 (samples/drivers/flash_shell/ or so)
to see the configuration when CM4 is accessing the flash.

mkaricheri commented:
> > @FRASTM @GeorgeCGV @ajarmouni-st CC: @ABESTM, @erwango I have addressed all of your comments and also rebased the branch to the latest on main and pushed. Please give your approval if this looks good.
> 
> I'd like please an example of the DTS for stm32h747_disco//m4 or stm32h745i_disco//m4 (samples/drivers/flash_shell/ or so) to see the configuration when CM4 is accessing the flash.

@FRASTM 747i-disco sample for flash+shell is already added. Please check. I will add for 745i-disco. I didn't realize 745i-disco is supported which was not available earlier,

mkaricheri committed:
samples: drivers: flash_shell: allow run on stm32h745i_disco/stm32h745xx/m4

Now that flash controller driver is supported on M4 of stm32h745i_disco,
remove it from platform_exclude list and add an overlay for flash_shell
sample.

Signed-off-by: Murali Karicheri <murali.karicheri@sandc.com>

mkaricheri commented:
@GeorgeCGV I have added support for 745i-disco per request from @FRASTM. So looks like your approval is gone. Please review and approve

mkaricheri commented:
@FRASTM @GeorgeCGV @ajarmouni-st CC: @ABESTM, @erwango, thanks for the review and getting this approved for merge. I assume this gets merged as per the merge window schedule, and there is nothing to be done from my side to get it merged. Please alert otherwise.
Thanks
Murali
